pkg/env/readers.go

```diff
@@ -116,13 +128,29 @@ func testReader(fallback cfg.Reader, overrider *testOverrides) cfg.Reader {
 // use the fName across two tests running in parallel. This will cause the
 // function to potentially panic.
 //
-// TODO[DT-3185]: Related work item to make the safety of this function better
+// Deprecated: Please use `FakeTestConfigHandler`
 func FakeTestConfig(fName string, ptr interface{}) func() {
 	// add ensures that it doesn't already exist to prevent two tests running
 	// concurrently colliding on fName.
-	overrides.add(fName, ptr)
-
+	err := overrides.addHandler(fName, ptr)
+	if err != nil {
+		panic(fmt.Sprintf("failed to addHandler '%v'. Please use the function 'FakeTestConfigHandler()' instead", err.Error()))
+	}
 	return func() {
 		overrides.delete(fName)
 	}
 }
+
+// FakeTestConfigHandler allows you to fake the test config with a specific value
+// and returns an error if a config with the same name exists already. If callers get an error,
+// they should switch to running tests in serial.
+func FakeTestConfigHandler(fName string, ptr interface{}) (func(), error) {
```

Comment:
In Go it's more idiomatic, when renaming a function signature, to rename it to indicate what has changed. For example, when adding a Context you'd do `func <Name>Context`.

With that in mind, I'd suggest naming this `FakeTestConfigWithError` or something along those lines. Otherwise, it's not very clear to the consumer what we're doing here.

We should also change `FakeTestConfig` to do _nothing_ but call `FakeTestConfigHandler` (as it is named today) instead of duplicating the logic.